pyqtgraph/widgets/MultiAxisPlotWidget.py

```diff
@@ -0,0 +1,378 @@
+            or the name of an axis previously created by calling addAxis().
+            This axis will be set in the new PlotItem based on the selected axis orientation.
+            If None "bottom" will be used as default.
+        y_axis (str, None):
+            one of the default PlotItem axis names ("top", "right", "bottom", "left"),
+            or the name of an axis previously created by calling addAxis().
+            This axis will be set in the new PlotItem based on the selected axis orientation.
+            If None "left" will be used as default.
+        chart (PlotDataItem, ...):
+            The chart to be used inside the new PlotItem.
+            If left as None a new PlotDataItem will be created and it's name parameter set.
+        *args and **kwargs:
+            Parameters to be passed to the newly created PlotItem.
+        Returns:
+            PlotDataItem: The newly created PlotDataItem.
+            PlotItem: The newly created PlotItem.
+        """
+        # add default axis to the list of axes if requested
+        if x_axis in ["bottom", "top"]:
+            x = self.addAxis(x_axis, axis=self.pi.axes[x_axis]["item"])
+        elif x_axis not in self.axes:
+            x = self.addAxis(x_axis)
+        else:
+            x = self.axes[x_axis]
+        if y_axis in ["left", "right"]:
+            y = self.addAxis(x_axis, axis=self.pi.axes[y_axis]["item"])
+        elif y_axis not in self.axes:
+            y = self.addAxis(y_axis)
+        else:
+            y = self.axes[y_axis]
+        if x_axis in ["bottom", "top"] and y_axis in ["left", "right"]:
+            # use default plotitem if none provided
+            plotitem = self.pi
+        else:
+            # VIEW
+            plotitem = PlotItem(parent=self.pi, name=name, *args, enableMenu=False, **kwargs)  # pass axisitems?
+            # hide all plotitem axis (they vould interfere with viewbox)
+            for a in ["left", "bottom", "right", "top"]:
+                plotitem.hideAxis(a)
+            # fix parent legend not showing child charts
+            plotitem.legend = self.pi.legend
+            for k, pos, axis in [["left", [2, 0], y], ["bottom", [3, 1], x],  ["right", [2, 2], y], ["top", [1, 1], x]]:
+                # # DO NOT USE, WILL MAKE AXIS UNMATCHED TO DATA
+                # # you can't add the new ones after, it doesn't work for some reason
+                # # hide them instead
+                # plotitem.layout.removeItem(plotitem.layout.itemAt(*pos))
+                if axis.orientation == k:
+                    plotitem.axes[k] = {"item": axis, "pos": pos}
+                # hide plotitem axes
+                plotitem.hideAxis(a)
+        # CHART
+        if chart is None:
+            chart = PlotDataItem(name=name)
+        plotitem.addItem(chart)
+        # keep plotitem inside chart
+        chart.plotItem = plotitem
+        # keeptrack of connections inside chart
+        chart.mapw_conn = {}
```

Comment:
Excellent work refactoring the signal stuff, but I'd like to push for an additional change. Motivation: adding novel attributes to **other** objects is usually a sign that encapsulation is being violated, which portends of problems with keeping things separated. Keeping track of all these signals+handlers should live entirely on `self`. At a minimum, this could be a nested dictionary on self: `self._signalHandlersByChart = {}`, and so on. Alternately, you might prefer to build a bridge class in between the MultiAxis and the individual PlotDataItems that is in charge of this and other glue logic.

Reply:
done the dict version